### ISSUE TITLE
Parts in score organization

### ISSUE THREAD

**Issue by davidrizo** [music-encoding/music-encoding#660]
Some time ago we were discussing in the MEI-L list [this issue](https://lists.uni-paderborn.de/pipermail/mei-l/2017/002058.html), but seems it has not been closed yet.

Let's take this example from the [sample encodings](https://github.com/music-encoding/sample-encodings/tree/master/MEI_4.0/Music/Complete_examples)
![image](https://user-images.githubusercontent.com/10772049/79011145-b52faa00-7b63-11ea-86b1-9694201870fe.png)

The first staff contains the "Soprano 1" part, the second staff contains two layers, one for the "Soprano 2" and the other for the "Alto", and the third staff with a layer for the "Tenor" and the other for the "Bass".

This is the excerpt of the [encoding](https://github.com/music-encoding/sample-encodings/blob/master/MEI_4.0/Music/Complete_examples/Altenburg_Macht_auf_die_Tor.mei):

* `/mei/meiHead/workList/work/perfMedium/perfResList/`
	* `perfRes`: Sopran 1
	* `perfRes`: Sopran 2
	* `perfRes`: Alt
	* `perfRes`: Tenor
	* `perfRes`: Bass
* `/mei/music/body/mdiv/score/scoreDef/staffGrp/`
	* `staffDef/label`: S1.
	* `staffDef/label`: S.2/A.
	* `staffDef/label`: T./B.
* Finally, notes are arranged into staves and layers (`/mei/music/body/mdiv/score/section/measure/staff/layer/note`) without any link to the different parts. 

With this information it's difficult to automatically extract the notes belonging to the tenor (one must figure-out which layer -either n=1 or n=2- in the staff corresponds to the tenor). The same applies both to the score mode (`mdiv / score`) and to the  _particella_ organization `mdiv / parts / part`.

## Workarounds
I've found two possible workarounds:
### Using instrDef
It's too tricky, but there is logical link between elements (layer or staff → instrument → perfRes).

```xml
...
<perfResList>  
	<perfRes xml:id="soprano1">Sopran 1</perfRes>  
	<perfRes xml:id="soprano2">Sopran 2</perfRes>  
	<perfRes xml:id="alto">Alt</perfRes>  
	<perfRes xml:id="tenor">Tenor</perfRes>  
	<perfRes xml:id="bass">Bass</perfRes>  
</perfResList>
...
<staffGrp symbol="bracket" bar.thru="false">  
	<staffDef n="1" clef.shape="G" clef.line="2" key.sig="0" lines="5">  
		<label>S1.</label>  
		<instrDef xml:id="instrS1" corresp="#soprano1"/>  
	</staffDef>
	...
	<staffDef n="3" clef.shape="F" clef.line="4" lines="5">  
		<staffDef n="3" clef.shape="F" clef.line="4" lines="5">  
		<label>T./B.</label>  
		<layerDef n="1">  
			<instrDef xml:id="instrT" corresp="#tenor"/>  
		</layerDef>  
		<layerDef n="1">  
			<instrDef xml:id="instrB" corresp="#bass"/>  
		</layerDef>  
	</staffDef>		
```

### Linking directly to perfRes:
It's easier but the `corresp` attribute may link to any other kind of element
```xml
...
<perfResList>  
	<perfRes xml:id="soprano1">Sopran 1</perfRes>  
	<perfRes xml:id="soprano2">Sopran 2</perfRes>  
	<perfRes xml:id="alto">Alt</perfRes>  
	<perfRes xml:id="tenor">Tenor</perfRes>  
	<perfRes xml:id="bass">Bass</perfRes>  
</perfResList>
...
<staffGrp symbol="bracket" bar.thru="false">  
	<staffDef n="1" clef.shape="G" clef.line="2" key.sig="0" lines="5" corresp="#soprano1">  
		<label>S1.</label>  
	</staffDef>
	...
	<staffDef n="3" clef.shape="F" clef.line="4" lines="5">  
		<label>T./B.</label>  
		<layerDef n="1" corresp="#tenor"/>  
		<layerDef n="2" corresp="#bass"/>  
	</staffDef>	
```
My proposal is:
- either to add a `partDef` containing a `partName` attribute anywhere to which the elements `perfRes`, `staffGrp`, `staffDef`, `layerDef` and part could be easily linked.
- or to have a new attribute `perfRes` in `layerDef`, `staffDef`, `staffGrp` and `part` to directly link to the part definition (I don't known if adding it also to `instrDef`)

**Comment by pe-ro** [comment]
Hi David,

What you propose is fine, but I suggest that if the goal is to link directly to a <perfRes> element, then the attribute should be named ‘perfRes’ and it should be required to take the value of an existing <perfRes> element’s @xml:id.

Another possibility is to use a value in @label <instrDef> that matches the corresponding <perfRes> element’s @xml:id attribute or its content.  For example:

<perfResList  xmlns="http://www.music-encoding.org/ns/mei">
  <perfRes xml:id="soprano1">Sopran 1</perfRes>
  <perfRes xml:id="soprano2">Sopran 2</perfRes>
  <perfRes xml:id="alto">Alt</perfRes>
  <perfRes xml:id="tenor">Tenor</perfRes>
  <perfRes xml:id="bass">Bass</perfRes>
</perfResList>
…
<staffGrp xmlns="http://www.music-encoding.org/ns/mei">
  <staffDef n="1" clef.shape="G" clef.line="2" key.sig="0" lines="5">
    <label>S1.</label>
    <instrDef xml:id="instrS1" label="soprano1"/>
    
    <instrDef xml:id="instrS1" label="Sopran 1"/>
  </staffDef>
  <staffGrp>
    <staffDef n="3" clef.shape="F" clef.line="4" lines="5">
      <label>T./B.</label>
      <layerDef n="1">
        <instrDef xml:id="instrT" label="tenor"/>
      </layerDef>
      <layerDef n="1">
        <instrDef xml:id="instrB" label="bass"/>
      </layerDef>
    </staffDef>
  </staffGrp>
</staffGrp>

This wouldn’t require adding a new attribute and could be enforced with schematron, although a mismatch between instrDef/@label and perfRes/@xml:id (or perfRes content) should probably generate a warning message, not a validation error.  This way instrDef/@label can still be “free-form”, but the encoder will get a helpful message.

The difference between the two options is in how strict the MEI schema should be and when the next version of MEI is expected to be released.

--
p.

**Comment by davidrizo** [comment]
Hi Perry,
I think the simplest option to understand is what you first suggest:
"...link directly to a <perfRes> element, then the attribute should be named ‘perfRes’ and it should be required to take the value of an existing <perfRes> element’s @xml:id."

Personally I don't need to have it now, it's just a proposal for improving MEI. 
If it's considered interesting for being added to the next version of MEI, the new attribute perfRes should be placed in `layerDef`, `staffDef`, `staffGrp` and part to directly link to the `part` definition (I don't known if adding it also to `instrDef`)

**Comment by davidrizo** [comment]
Maybe, it makes more sense to use directly `instrDef` rather than the `perfRes`. Besides, we can link directly the `layerDef` to the `instrDef` in the current MEI version:

```xml
<scoreDef>
                        <instrGrp>
                            <instrDef xml:id="soprano" label="Soprano"></instrDef>
                            <instrDef xml:id="alto" label="Alto"></instrDef>
                            <instrDef xml:id="tenor" label="Tenor"></instrDef>
                            <instrDef xml:id="bass" label="Bass"></instrDef>
                        </instrGrp>
                        <staffGrp>
                            <staffDef n="1" lines="5" key.sig="0" key.mode="major" meter.sym="common" clef.shape="G" clef.line="2">
                                <layerDef n="1" instr="#soprano"/>
                                <layerDef n="2" instr="#alto"/>
                            </staffDef>
                            <staffDef n="2" lines="5" key.sig="0" key.mode="major" meter.sym="common" clef.shape="F" clef.line="4">
                                <layerDef n="1" instr="#soprano"/>
                                <layerDef n="2" instr="#bass"/>                                
                            </staffDef>
                        </staffGrp>
                    </scoreDef>
```

**Comment by pe-ro** [comment]
Yep, that works too.  Definitely makes more sense, as it separates the info needed to associate a layer with an instrument from the textual label / metadata for the performing forces.  Sorry I didn’t recall this before.
